From: Zhen Wei <zwei@novell.com>
Subject: ocfs2: outstanding scheduled work can oops when quorum is shut down
Patch-mainline: Never - userspace cluster patch-dependent
References: 220694

 If there is outstanding work when o2quo_exit() is called, the work can oops
 when it is completed.

 This patch cancels the delayed work before completing the shutdown.
 
Acked-by: Jeff Mahoney <jeffm@suse.com>
diff -uprN /a/fs/ocfs2/cluster/quorum.c b/fs/ocfs2/cluster/quorum.c
--- a/fs/ocfs2/cluster/quorum.c	2006-11-14 14:24:33.000000000 +0800
+++ b/fs/ocfs2/cluster/quorum.c	2006-11-14 14:21:54.000000000 +0800
@@ -392,6 +392,11 @@ int o2quo_init(void)
 
 void o2quo_exit(void)
 {
+	struct o2quo_state *qs = &o2quo_state;
+	long i;
+	for (i = 0; i < O2NM_MAX_NODES; i++)
+		cancel_delayed_work(&qs->qs_node_work[i]);
+
 	flush_scheduled_work();
 	o2quo_unregister_hb_callbacks();
 	o2net_unregister_notifier(&o2quo_net_nb);
